velox/exec/StreamingWindow.cpp

```diff
@@ -0,0 +1,199 @@
+      // wholeGroupedRows_ into the sortedRows and then update the
+      // partitionStartRows_ value.
+      updateValues();
+    }
+
+    wholeGroupedRows_.push_back(newRow);
+    preRow_ = newRow;
+  }
+
+  partitionStartRows_.push_back(sortedRows_.size());
+
+  outputBatchSize_ = outputBatchRows(data_->estimateRowSize());
+}
+
+inline bool StreamingWindow::compareRowsWithKeys(
+    const char* lhs,
+    const char* rhs,
+    const std::vector<std::pair<column_index_t, core::SortOrder>>& keys) {
+  if (lhs == rhs) {
+    return false;
+  }
+  for (auto& key : keys) {
+    if (auto result = data_->compare(
+            lhs,
+            rhs,
+            key.first,
+            {key.second.isNullsFirst(), key.second.isAscending(), false})) {
+      return result < 0;
+    }
+  }
+  return false;
+}
+
+void StreamingWindow::createPeerAndFrameBuffers() {
+  // TODO: This computation needs to be revised. It only takes into account
+  // the input columns size. We need to also account for the output columns.
+  // numRowsPerOutput_ = outputBatchRows(data_->estimateRowSize());
+
+  numRowsPerOutput_ = numRows_;
+
+  peerStartBuffer_ = AlignedBuffer::allocate<vector_size_t>(
+      numRowsPerOutput_, operatorCtx_->pool());
+  peerEndBuffer_ = AlignedBuffer::allocate<vector_size_t>(
+      numRowsPerOutput_, operatorCtx_->pool());
+
+  auto numFuncs = windowFunctions_.size();
+
+  frameStartBuffers_.clear();
+  frameEndBuffers_.clear();
+  validFrames_.clear();
+  frameStartBuffers_.reserve(numFuncs);
+  frameEndBuffers_.reserve(numFuncs);
+  validFrames_.reserve(numFuncs);
+
+  for (auto i = 0; i < numFuncs; i++) {
+    BufferPtr frameStartBuffer = AlignedBuffer::allocate<vector_size_t>(
+        numRowsPerOutput_, operatorCtx_->pool());
+    BufferPtr frameEndBuffer = AlignedBuffer::allocate<vector_size_t>(
+        numRowsPerOutput_, operatorCtx_->pool());
+    frameStartBuffers_.push_back(frameStartBuffer);
+    frameEndBuffers_.push_back(frameEndBuffer);
+    validFrames_.push_back(SelectivityVector(numRowsPerOutput_));
+  }
+}
+
+void StreamingWindow::noMoreInput() {
+  Operator::noMoreInput();
+}
+
+bool StreamingWindow::isFinished() {
+  return noMoreInput_ && data_->numRows() == 0 && output_ == nullptr;
+}
+
+RowVectorPtr StreamingWindow::createFinalResult() {
+  if (output_->size() > outputBatchSize_) {
+    auto result = std::dynamic_pointer_cast<RowVector>(
+        output_->slice(0, outputBatchSize_));
+    auto remainingSize = output_->size() - outputBatchSize_;
+    output_ = std::dynamic_pointer_cast<RowVector>(
+        output_->slice(outputBatchSize_, remainingSize));
+    return result;
+  } else if (noMoreInput_) {
+    auto result = std::move(output_);
+    return result;
+  } else {
+    return nullptr;
+  }
+}
+
+RowVectorPtr StreamingWindow::createOutput() {
+  numRows_ = sortedRows_.size();
+
+  currentPartition_ = 0;
+  numProcessedRows_ = 0;
+
+  createPeerAndFrameBuffers();
```

Comment:
Why is this being done for each output call ? It might be better to allocate those once and only re-allocate each time if needed.

Reply:
My current implementation processes only the first n-1 partitions for each input, while the last partition is kept for computation with the next input because it may contain data from the same partition as the next input. As a result, the number of rows for each output calculation is not fixed. Therefore, it is necessary to calculate it every time the output is called. I have updated the buffer as re-usable as much as possible in createPeerAndFrameBuffers method.